### DISCUSSION TITLE
Multiple custom_targets with the same name cause issues

### DISCUSSION THREAD

**Discussion by akbyrd** [mesonbuild/meson#12525] [Q&A]
I run a code generation step for each source file in my build. The generated file name is `@BASENAME@.generated.cpp`. This causes a problem when there are two files in different directories with the same name because all generated files are put at the root of the build directory.

Source Files
```
- foo.cpp
  - sub/foo.cpp
```

Build File
```
project('code-gen-test', 'cpp')

source_files = files('foo.cpp', 'sub/foo.cpp')

generator         = files('code-gen.ps1')
generated_targets = []

foreach source_file : source_files
	generated_target = custom_target(
		'@0@'.format(source_file).replace('/', '-'),
		input: source_file,
		output: [ '@BASENAME@.generated.cpp' ],
		command: [ 'pwsh', '-File', generator, '@INPUT@', '@OUTPUT@' ],
		depend_files: generator
	)
	generated_targets += generated_target
endforeach

executable('code-gen-test', source_files + generated_targets)
```

What's the proper way to handle this? I was hoping to end up with
```
- build/
  - foo.generated.cpp
  - sub/
    - foo.generated.cpp
- src/
  - foo.cpp
  - sub/
    - foo.cpp
```

**Comment by mitchgrout** [comment]
One option would be to rewrite the `custom_target` as a `generator` instead, and invoke `generator.process` in a `subdir` to ensure you place your files into mirrored directories in the build directory.

I personally feel that there should be some way to do what you've got _without_ having to invent a `subdir()` just to ensure the build directory is happy.

**Reply by akbyrd** [reply]
That implies the need for a `meson.build` in every subdirectory in order to avoid name clashes.

The directory structure I see most often for a C++ game engine + project (the context I'm working in) is roughly
```
engine/
  src/
    SystemA/
    SystemB/
    SystemC/
game/
  src/
    FeatureX/
    FeatureY/
    FeatureZ/
```

Name clashes are rare, but they do happen. Most often with something like `FeatureX/Types.h` and `FeatureY/Types.h`. Name clashes between the engine and the project aren't an issue, since I expect each to have their own build file and therefore subdirectory in the build directory. There's a reasonable argument for making all file names unique, and it's something I generally aim for anyway for other reasons, but it's a difficult sell when sharing that engine (plus build process) across multiple games/teams/users. There are other reasons why I'd argue that generated code should mirror the directory structure of the files that caused it.
